### PR TITLE
Cryp 144/unittests account creation

### PR DESCRIPTION
**Pull request by aafshar-poly**
An attempt in adding mercat account creation and validation to the chain. I have a couple of questions about it which I have marked as TODOs. I'll also make a github comment where I have questions :)

### CODE REVIEW

File: pallets/confidential/src/lib.rs [hunk 1]
```diff
@@ -79,6 +88,36 @@ decl_module! {
 
         fn deposit_event() = default;
 
+        #[weight = 400_000]
+        pub fn store_valid_asset_ids(origin, // TODO: how to ensure only privileged users can call this?
```
Comment (aafshar-poly) [inline pallets/confidential/src/lib.rs:92]: We need a place on the chain to store the list of valid asset ids (encrypted form of ticker names). How can I enforce that only authorized entities can call and update that list.

File: pallets/confidential/src/lib.rs [hunk 2]
```diff
@@ -79,6 +88,36 @@ decl_module! {
 
         fn deposit_event() = default;
 
+        #[weight = 400_000]
```
Comment (aafshar-poly) [inline pallets/confidential/src/lib.rs:91]: Weights are to be decided. Using 400,000 for everything for now.

File: pallets/confidential/src/lib.rs [hunk 3]
```diff
@@ -79,6 +88,36 @@ decl_module! {
...
+        pub fn store_account_tx(origin,
+            tx: PubAccountTx,
+        ) -> DispatchResult
+        {
+            let owner_acc = ensure_signed(origin)?;
+            let owner_id = Context::current_identity_or::<Identity<T>>(&owner_acc)?;
+
+            let account_vldtr = AccountValidator {};
+            let valid_asset_ids = Self::valid_asset_ids();
+            let valid_asset_ids = convert_asset_ids(valid_asset_ids);
```
Comment (aafshar-poly) [inline pallets/confidential/src/lib.rs:112]: Right now I am calling `convert_asset_ids` to convert them from `AssetId` type to `Scalar` type.
- One reason was that storing `Scalar` on the chain would not work since `Scalar` does not implement Encode/Decode.
- The other reason is that I am not yet sure about how we will end up converting plain ticker names to asset ids.

File: pallets/confidential/src/lib.rs [hunk 4]
```diff
@@ -70,6 +75,10 @@ decl_storage! {
         pub RangeProofs get(fn range_proof): double_map hasher(twox_64_concat) IdentityId, hasher(blake2_128_concat) ProverTickerKey => Option<TickerRangeProof>;
 
         pub RangeProofVerifications get(fn range_proof_verification): double_map hasher(blake2_128_concat) (IdentityId, Ticker), hasher(twox_64_concat) IdentityId => bool;
+
+        pub MercatAccounts get(fn mercat_accounts): map hasher(blake2_128_concat) IdentityId => Vec<PubAccount>;
```
Comment (aafshar-poly) [inline pallets/confidential/src/lib.rs:79]: This data structure is supposed to hold the list of all the meract accounts that an identity holds. Is this the right place for storing it? Or should be be moved to the `asset` pallet?

File: pallets/runtime/tests/src/confidential_test.rs [hunk 5]
```diff
@@ -14,6 +14,17 @@ use test_client::AccountKeyring;
...
+    asset_proofs::{CommitmentWitness, ElgamalSecretKey},
+    mercat::{
+        account::{convert_asset_ids, AccountCreator},
+        AccountCreatorInitializer, EncryptionKeys, PubAccountTx, SecAccount,
+    },
+    AssetId,
+};
+use curve25519_dalek::scalar::Scalar;
+use rand::{rngs::StdRng, SeedableRng};
+use schnorrkel::{ExpansionMode, MiniSecretKey};
```
Comment (aafshar-poly) [inline pallets/runtime/tests/src/confidential_test.rs:27]: In the final version of this PR, these two will most likely be removed. More on this in the comment below.

File: pallets/runtime/tests/src/confidential_test.rs [hunk 6]
```diff
@@ -73,3 +84,59 @@ fn range_proof_we() {
...
+
+fn account_create_tx_we() {
+    let alice = AccountKeyring::Alice.public();
+    let alice_id = register_keyring_account(AccountKeyring::Alice).unwrap();
+
+    let valid_asset_ids: Vec<AssetId> = vec![1, 2, 3]
+        .iter()
+        .map(|id| AssetId::from(id.clone()))
+        .collect();
+    Confidential::store_valid_asset_ids(Origin::signed(alice), valid_asset_ids.clone());
```
Comment (aafshar-poly) [inline pallets/runtime/tests/src/confidential_test.rs:103]: These asset ids should NOT be updated by Alice.

File: pallets/runtime/tests/src/confidential_test.rs [hunk 7]
```diff
@@ -73,3 +84,59 @@ fn range_proof_we() {
...
+    let elg_secret = ElgamalSecretKey::new(Scalar::random(&mut rng));
+    let elg_pub = elg_secret.get_public_key();
+    let enc_keys = EncryptionKeys {
+        pblc: elg_pub.into(),
+        scrt: elg_secret.into(),
+    };
+
+    let sign_keys = MiniSecretKey::from_bytes(&[11u8; 32])
+        .expect("Invalid seed")
+        .expand_to_keypair(ExpansionMode::Ed25519);
```
Comment (aafshar-poly) [inline pallets/runtime/tests/src/confidential_test.rs:115]: Is it ok for the tests to use rand and generate secrets? Or do you have a different way of mocking the Wallet's behaviour?

File: pallets/runtime/tests/src/confidential_test.rs [hunk 8]
```diff
@@ -73,3 +84,59 @@ fn range_proof_we() {
...
+    ExtBuilder::default()
+        .build()
+        .execute_with(account_create_tx_we);
+}
+
+fn account_create_tx_we() {
+    let alice = AccountKeyring::Alice.public();
+    let alice_id = register_keyring_account(AccountKeyring::Alice).unwrap();
+
+    let valid_asset_ids: Vec<AssetId> = vec![1, 2, 3]
```
Comment (aafshar-poly) [inline pallets/runtime/tests/src/confidential_test.rs:99]: `[1, 2, 3]` is supposed to represent the plain ticker ids. Is this a fair representation/assumption?

File: primitives/src/cdd_id.rs [hunk 9]
```diff
@@ -49,8 +50,8 @@ impl CddId {
     /// and `investor_uid`.
     pub fn new(did: IdentityId, investor_uid: InvestorUid) -> Self {
         let cdd_claim_data = CDDClaimData {
-            investor_did: RawData(did.to_bytes()),
-            investor_unique_id: RawData(investor_uid.0),
+            investor_did: Scalar::from_bits(did.to_bytes()),
+            investor_unique_id: Scalar::from_bits(investor_uid.0),
         };
```
Comment (aafshar-poly) [inline primitives/src/cdd_id.rs:55]: These changes are unrelated to the topic of this PR. The cryptography library and Polymesh were out of sync and I needed to do this to make the code compile.

Comment (ParnianAlimi) [reply primitives/src/cdd_id.rs:55]: It might have been resolved after https://github.com/PolymathNetwork/Polymesh/pull/499 went in, did you rebase on the latest `develop`?

File: pallets/confidential/src/lib.rs (hunk 1 above)
Comment (Centril) [reply pallets/confidential/src/lib.rs:92]: Sounds like you need something like:
```rust
T::CommitteeOrigin::ensure_origin(origin)?;
```
(see e.g., pips pallet)

File: pallets/confidential/src/lib.rs [hunk 10]
```diff
@@ -79,6 +88,36 @@ decl_module! {
...
+            let owner_id = Context::current_identity_or::<Identity<T>>(&owner_acc)?;
+
+            let account_vldtr = AccountValidator {};
+            let valid_asset_ids = Self::valid_asset_ids();
+            let valid_asset_ids = convert_asset_ids(valid_asset_ids);
+            let _ = account_vldtr.verify(&tx, &valid_asset_ids).map_err(|_| Error::<T>::InvalidAccountCreationProof)?;
+
+            let mut mercat_accounts = Self::mercat_accounts(owner_id);
+            mercat_accounts.push(tx.pub_account);
+            <MercatAccounts>::insert(&owner_id, mercat_accounts);
```
Comment (Centril) [inline pallets/confidential/src/lib.rs:117]:
```suggestion
            <MercatAccounts>::append(&owner_id, tx.pub_account);
```
(+ remove 2 lines before)

File: pallets/confidential/src/lib.rs (hunk 2 above)
Comment (satyamakgec) [reply pallets/confidential/src/lib.rs:91]: You can put weigh in a multiple of 1_000_000_000

File: pallets/confidential/src/lib.rs (hunk 3 above)
Comment (satyamakgec) [reply pallets/confidential/src/lib.rs:112]: I am assuming the asset_ids are unique and encrypted in some sense so no one can know which asset ticker has which asset_id.

FYI - We can use the asset Identity as well that is generated on the basis of the hash(SecurityToken + Ticker) if assetId is not encrypted.

File: pallets/confidential/src/lib.rs (hunk 4 above)
Comment (satyamakgec) [reply pallets/confidential/src/lib.rs:79]: I am more inclined to use the different pallets i.e `confidential-asset`. We can discuss more about this if you let me know the expected API for the confidential asset.

File: pallets/runtime/tests/src/confidential_test.rs (hunk 6 above)
Comment (satyamakgec) [reply pallets/runtime/tests/src/confidential_test.rs:103]: I am assuming this is only for the test

File: pallets/runtime/tests/src/confidential_test.rs (hunk 7 above)
Comment (satyamakgec) [reply pallets/runtime/tests/src/confidential_test.rs:115]: In general, we are using `substrate-test-utils` pallet which will allow us to use Alice , Bob, Charlie, or different test accounts which allow you to sign the transaction. But you have to check whether it gives you all the utility or not. 

I am not sure what you are referring to generate secrets, If it is a key pair then substrate-test-utils is the best fit.
using rand is fine in test

File: pallets/runtime/tests/src/confidential_test.rs (hunk 8 above)
Comment (satyamakgec) [reply pallets/runtime/tests/src/confidential_test.rs:99]: Yeah it is fine

File: pallets/confidential/src/lib.rs (hunk 4 above)
Comment (satyamakgec) [inline pallets/confidential/src/lib.rs:79]: Can we hold something like that 
```
pub MercatAccounts get(fn mercat_accounts): double_map hasher(blake2_128_concat) asset_id, hasher(blake2_128_concat) IdentityId => PubAccount;
```
It will allow accessing the list of all identity who has the given asset id easily and also it will save us from encoding/decoding of a vector on-chain when we need to add/delete the account or access the account for a given asset_id

File: pallets/confidential/src/lib.rs [hunk 11]
```diff
@@ -79,6 +88,36 @@ decl_module! {
...
+            tx: PubAccountTx,
+        ) -> DispatchResult
+        {
+            let owner_acc = ensure_signed(origin)?;
+            let owner_id = Context::current_identity_or::<Identity<T>>(&owner_acc)?;
+
+            let account_vldtr = AccountValidator {};
+            let valid_asset_ids = Self::valid_asset_ids();
+            let valid_asset_ids = convert_asset_ids(valid_asset_ids);
+            let _ = account_vldtr.verify(&tx, &valid_asset_ids).map_err(|_| Error::<T>::InvalidAccountCreationProof)?;
```
Comment (satyamakgec) [inline pallets/confidential/src/lib.rs:113]: I think you can use the associated function something like that 
```
AccountValidator::verify(&tx, &valid_asset_ids).map_err(|_| Error::<T>::InvalidAccountCreationProof)?;
```

File: pallets/confidential/src/lib.rs (hunk 1 above)
Comment (aafshar-poly) [reply pallets/confidential/src/lib.rs:92]: Thanks @Centril

File: pallets/confidential/src/lib.rs (hunk 3 above)
Comment (aafshar-poly) [reply pallets/confidential/src/lib.rs:112]: Yes, the asset_ids of MERACT lib are encrypted and unique. Great, asset identity works well.

File: pallets/confidential/src/lib.rs (hunk 4 above)
Comment (aafshar-poly) [reply pallets/confidential/src/lib.rs:79]: We'll be spending some time in the next couple of days figuring out the API and the storage that we need on the Polymesh. We can revisit this discussion once that is done.

File: pallets/runtime/tests/src/confidential_test.rs (hunk 6 above)
Comment (aafshar-poly) [reply pallets/runtime/tests/src/confidential_test.rs:103]: Yep, this was kinda a negative test ;)

**Comment by aafshar-poly** [comment]
Thanks for the feedback guys. I'll close this PR, rebase it on the new "mercat" branch and submit another PR next week.
